Autodetect parse() format

dwinston opened the pull request:
Addresses #981

* Use `rdflib.util.guess_format` to guess and try to apply format based on file extension
* Log warning if assumed XML(not guessed from file extension) and cannot parse
* Add tests

Close #981

tgbugs commented on rdflib/graph.py:
```diff
@@ -1070,13 +1072,24 @@ def parse(
-            # "expicitly specify one with the format argument." % source)
-            format = "application/rdf+xml"
+            try:
+                from rdflib.util import guess_format  # local import avoids circular dependency
```
It is probably better to just move the import to the end of the file but keep it at the top level in this situation.

dwinston replied:
I refactored to use an absolute module import at the top, as per [this suggestion](https://stackoverflow.com/a/37126790/1820042). I changed two imports in `rdflib.util` to accommodate this. This seems like an improvement.

tgbugs commented on rdflib/graph.py:
```diff
@@ -1070,13 +1072,24 @@ def parse(
+            try:
+                from rdflib.util import guess_format  # local import avoids circular dependency
+                format = guess_format(source.file.name)
+            finally:
```
It is not clear why this is wrapped in a try/finally block. Any error that occurs inside the try block will be fatal and parsing will not be able to continue.

dwinston replied:
Thanks for pointing this out. I somehow brain-farted on the semantics of try/except/finally. I pushed what should hopefully fix this -- I catch specific exceptions that are to be expected.

dwinston commented:
@tgbugs what do you thing? I hope I've resolved your concerns.

tgbugs commented on rdflib/graph.py:
```diff
@@ -1070,13 +1073,24 @@ def parse(
-            # "expicitly specify one with the format argument." % source)
-            format = "application/rdf+xml"
+            try:
+                format = rdflib.util.guess_format(source.file.name)
```
Under what circumstances are you seeing these errors? I'm guessing when `source.file.name is None` for the TypeError and probably ... when `not hasattr(source, 'file')`? If that is the case then an explicit test (if statement) for the presence and non-Noneness of the values would be much clearer here.

dwinston replied:
@tgbugs I've adopted your suggestion for more of a look-before-you-leap approach with an explicit if-statement guard, ensuring that `source.file.name` is an accessible path and is a string before attempting a call to `guess_format`.

nicholascar commented:
@dwinston please merge in `master` branch and address conflicts since a whole host of changes were recently merged in to remove Python 2 code. This will trigger a re-checking of the code by Travis. Then we can redress the comments from @tgbugs

dwinston commented:
> @dwinston please merge in `master` branch and address conflicts since a whole host of changes were recently merged in to remove Python 2 code. This will trigger a re-checking of the code by Travis. Then we can redress the comments from @tgbugs

Sure, will do.

dwinston commented:
@nicholascar merged in master, and I hope I addressed the comments from @tgbugs with my last commit. Also, I thought I'd note here that the more recent PR #1064 is an alternative solution to the same underlying issue.
